add `ZarrStore.getRaw`

manzt opened the pull request:
Adds feature #15 !

Ok, so this was a bit of a doozy, but things are working! 

I created a `RawArray` interface for writing to a single _strided_ `TypedArray` from "C" ordering. The following API is exposed to the end user:

```javascript
import { openArray, slice } from 'zarr';

const config = {
  store: 'http://localhost:5000',
  path: 'dummy_data.zarr',
};

const selection = [
    null, 
    slice(null, 512), 
    slice(null, null, 2)
];

const z = await openArray(config);
const rawArray = await z.getRaw(selection [, getOptions ]);

const { data, shape, strides } = rawArray; // data is single TypedArray

const arr = tf.tensor(data, shape); // tensorflow.js
const arr2 = ndarray(data, shape) ; // ndarray
```

## Technical details

When indexing the `ZarrStore`, `getRaw` works by creating a single `RawArray` as output. 

For each projection of the `indexer`,

- Entire chunk is fetched and decoded (like in `.get`), 
- The destination selection is used to index the output array while the chunk selection is used to index the `TypedArray` chunk to fill the output, rather than making a temporary array.

The final result is a `RawArray` with properties:
- `data`: Single strided `TypedArray` (e.g. `Int32Array(786432) [ 0, 1, 2, … ]`)
- `dtype`: Datatype string from zarr (e.g. `"<i4"`)
- `shape`: Array of dimension sizes (e.g. `[3, 512, 512]`)
- `strides`: Number of elements to jump per dimension (e.g. `[262144, 512, 1]`)

## Notes
- ~~One optimization that I haven't implemented yet is adding a check for whether a chunk _is_ exactly the desired selection. Rather than copying that entire chunk to the output, we should just directly return the decoded chunk (since it is already strided due to zarr).~~
- This is something we are interested in particularly with tiling image data. Tile-sizes === chunk sizes in our image pyramids, and this optimization would effectively make 1 request per tile 👍 
- Edit: I have now added this

manzt commented:
Here is a repo where I've been comparing outputs from `NestedArray` and compatibility with `tfjs`.

https://github.com/manzt/zarr-testing

manzt commented:
Motivating example;

```javascript
import { openArray, slice } from 'https://cdn.pika.dev/zarr/^0.1.5';

const config = {
  store: "http://localhost:5000",
  path: "image_tiles.zarr"
}; 
// shape:  [3, 2048, 2048]
// chunks: [3, 512, 512]

const selection = [
  null,          // image channels
  slice(0, 512), // y coordinates
  slice(0, 512), // x coords
];

const z = await openArray(config);

// fetches single chunk (tile) and returns RawArray directly
const { data, strides, shape } = await z.getRaw(selection);

// create image channel views
const imageSize = shape[1] * shape[2]; // 512 x 512
const r = new Float32Array(data.buffer, data.BYTES_PER_ELEMENT * strides[0] * 0, imageSize);
const g = new Float32Array(data.buffer, data.BYTES_PER_ELEMENT * strides[0] * 1, imageSize);
const b = new Float32Array(data.buffer, data.BYTES_PER_ELEMENT * strides[0] * 2, imageSize);

// pass to renderer...
```

ilan-gold commented:
I am not a `zarr` user beyond this use-case myself, but this strikes me as something that people would even expect, if they are using the `chunk_size` parameter like I am.

ilan-gold commented:
Thanks @manzt!

manzt commented:
Sorry for the updates, but I switched the algorithm to using `TypedArray` methods where possible. There is a performance benefit to this, and fortunately `TypedArray.subarray` offers the ability to pass a new "sub view" of the underlying `ArrayBuffer`; now it very closely reflects the logic used in `nestedArray` and is easier to read IMO.

gzuidhof commented on src/core/index.ts:
```diff
@@ -257,13 +258,19 @@ export class ZarrArray {
   public get(selection?: ArraySelection, opts?: GetOptions): Promise<NestedArray<TypedArray> | number>;
-  public get(selection: ArraySelection = null, opts: GetOptions = {}): Promise<NestedArray<TypedArray> | number> {
-    return this.getBasicSelection(selection, opts);
+  public get(selection: ArraySelection = null, opts: GetOptions = {}): Promise<NestedArray<TypedArray> | RawArray | number> {
```
I don't think this can ever return a RawArray as we pass `false`, maybe we can change the typing of `getBasicSelection` to return a different type given the input. I'll have a stab at it in a comment below

If it doesn't end up really possible to type it correctly, we can consider casting it here

manzt replied:
fixed! thanks

gzuidhof commented on src/core/index.ts:
```diff
@@ -257,13 +258,19 @@ export class ZarrArray {
-  public async getBasicSelection(selection: ArraySelection, { concurrencyLimit = 10, progressCallback }: GetOptions = {}): Promise<number | NestedArray<TypedArray>> {
+  public getRaw(selection?: undefined | Slice | ":" | "..." | null | (Slice | null | ":" | "...")[], opts?: GetOptions): Promise<RawArray>;
+  public getRaw(selection?: ArraySelection, opts?: GetOptions): Promise<RawArray | number>;
+  public getRaw(selection: ArraySelection = null, opts: GetOptions = {}): Promise<NestedArray<TypedArray> | RawArray | number> {
```
same here

gzuidhof commented on src/core/index.ts:
```diff
@@ -257,13 +258,19 @@ export class ZarrArray {
+
+  public async getBasicSelection(selection: Slice | ":" | "..." | null | (Slice | null | ":" | "...")[], asRaw?: boolean, opts?: GetOptions): Promise<NestedArray<TypedArray> | RawArray>;
+  public async getBasicSelection(selection: ArraySelection, asRaw?: boolean, opts?: GetOptions): Promise<NestedArray<TypedArray> | number | RawArray>;
+  public async getBasicSelection(selection: ArraySelection, asRaw = false, { concurrencyLimit = 10, progressCallback }: GetOptions = {}): Promise<number | NestedArray<TypedArray> | RawArray> {
```
my attempt at typing this (without an editor to verify):

```typescript
// asRaw = false
public async getBasicSelection(selection: Slice | ":" | "..." | null | (Slice | null | ":" | "...")[], asRaw?: false, opts?: GetOptions): Promise<NestedArray<TypedArray>>;
public async getBasicSelection(selection: ArraySelection, asRaw?: false, opts?: GetOptions): Promise<NestedArray<TypedArray> | number>;
// asRaw= true
public async getBasicSelection(selection: Slice | ":" | "..." | null | (Slice | null | ":" | "...")[], asRaw?: true, opts?: GetOptions): Promise<RawArray>;
public async getBasicSelection(selection: ArraySelection, asRaw?: true, opts?: GetOptions): Promise<number | RawArray>;
// Altogether
public async getBasicSelection(selection: ArraySelection, asRaw = false, { concurrencyLimit = 10, progressCallback }: GetOptions = {}): Promise<number | NestedArray<TypedArray> | RawArray> {
```

manzt replied:
This is my first project with Typescript 😬 so I really appreciate the feedback! Let me take a stab at this.

gzuidhof commented on src/core/index.ts:
```diff
@@ -396,6 +427,10 @@ export class ZarrArray {
     return new DTYPE_TYPEDARRAY_MAPPING[this.dtype](buffer);
   }
 
+  private toRawArray(buffer: Buffer | ArrayBuffer) {
```
I was a bit confused at first how we can just assume the chunksize or dtype here, but we only ever call this with a whole chunk, then it makes sense. Maybe this can be `chunkToRawArray`? or `chunkBufferToRawArray` or something like that?

manzt replied:
Good idea. I've been thinking of changing the `out.set()` function within `ZarrArray.getChunkItem` to `out.setFromChunkItem` because it's a very specific task of decoding a full chunk (with its selection) directly to the output array (which I doubt users will be trying to do directly). 

This seems distinct from what would be expected with `RawArray.set(typedArray, selection)`, where the typed array must be the exact size of the selection, and the indexing of `typedArray` can be inferred from the selection itself.

gzuidhof commented on src/rawArray/index.ts:
```diff
@@ -0,0 +1,94 @@
+
+    constructor(data: TypedArray, shape?: number | number[], dtype?: DtypeString, strides?: number[])
+    constructor(data: Buffer | ArrayBuffer | null, shape?: number | number[], dtype?: DtypeString, strides?: number[])
+    constructor(data: Buffer | ArrayBuffer | TypedArray | null, shape?: number | number[], dtype?: DtypeString, strides?: number[]) {
```
Note for myself: one day this and the nestedarray constructor should be cleaned up

manzt replied:
I'll add an issue!

gzuidhof commented on src/rawArray/ops.ts:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        for (let i = 0; i < outputSize; i++) {
+            dstArr[currentDstStride * (from + (step * i))] = sourceArr[currentSourceStride * (sfrom + (sstep * i))];
```
Hats off for getting this correct, it took me a while for NestedArrays and I think those were actually easier

manzt replied:
There was a bit of mental overhead with figuring out what you implemented. But thanks for that! No way I would have been able to otherwise probably, and I'm quite happy with the final result :) Learned a lot about `TypedArray`s and views of the underlying data buffers.

gzuidhof commented on src/rawArray/ops.ts:
```diff
@@ -0,0 +1,114 @@
+            nextDstSliceIndicies,
+        );
+    }
+}
```
nit: newline

gzuidhof commented on src/rawArray/ops.ts:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    // Get current indicies and strides for both destination and source arrays
+    const [currentDstSlice, ...nextDstSliceIndicies] = dstSliceIndices;
```
Nice, I think I just used `x = list[0]` and `xNext = list.slice(1)`, but this is much nicer

manzt replied:
Yeah, it took me some time to figure out what was going on, so I used destructuring and long variable names to help me remember what was going on 😅

gzuidhof commented on src/rawArray/ops.ts:
```diff
@@ -0,0 +1,114 @@
+        for (let i = 0; i < outputSize; i++) {
+            dstArr[currentDstStride * (from + (step * i))] = value;
+        }
+        return;
```
nit: I personally prefer using `else` in this case, I find it easier to read as it's more obviously "either this or that happens".

manzt replied:
Agreed. It was that way in NestedArray, so I used the same. Changed there and here!

gzuidhof commented on src/util.ts:
```diff
@@ -180,4 +180,22 @@ export function arrayEquals1D(a: ArrayLike<any>, b: ArrayLike<any>) {
+ *
+ * NOTE: These strides here are distinct from numpy.ndarray.strides, which describe actual byte steps.
+ */
+export function getStrides(shape: number[]): number[] {
```
👏

gzuidhof commented on src/util.ts:
```diff
@@ -180,4 +180,22 @@ export function arrayEquals1D(a: ArrayLike<any>, b: ArrayLike<any>) {
+    // adapted from https://github.com/scijs/ndarray/blob/master/ndarray.js#L326-L330
+    const ndim = shape.length;
+    const strides = Array(ndim);
+    let step = 1; // strides assumed to be contiguous, so init step is 1
```
I would propose to move this comment to the function documentation. If we ever start playing with views this assumption may not be true anymore

gzuidhof commented on test/rawArray/ops.setting.test.ts:
```diff
@@ -0,0 +1,168 @@
+            t.sourceShape,
+            t.sourceSelection
+        );
+        if (t.name === "4d_1x2x2x4_step_2") {
```
I think you didn't mean to leave this :)

gzuidhof commented on test/rawArray/ops.setting.test.ts:
```diff
@@ -0,0 +1,168 @@
+        expect(destRawArray.data).toEqual(t.expected);
+    });
+
+});
```
It would be really great if we could also test the end to end example where we call `getRaw` on a zarr array. 

For instance we could do it here by having another table driven test for getRaw too:
https://github.com/gzuidhof/zarr.js/blob/master/test/core/indexing.test.ts#L226-L230 (and also above for the 1D case)

gzuidhof commented:
If it's ok, I will send you an e-mail in a bit (maybe tomorrow), I'd love to get your input on a small other project I'm working on that's not (yet) open source

manzt commented:
I'm really happy with how this turned out! I added end-to-end tests for `getRaw` on a `ZarrArray`. These tests are now shared with `NestedArray` so the whole suite will be tested. In addition, I added some new signatures to `RawArray.set()`:

```javascript
    public set(selection: ArraySelection, value: RawArray | number): void;
    public set(selection: ArraySelection, chunk: RawArray, chunkSelection: ArraySelection): void;
    public set(selection: ArraySelection = null, value: RawArray | number, chunkSelection?: ArraySelection) 
```

The top signature is used the same exact way as `NestedArray`. If `value` is `number`, the `setRawArrayToScalar` method is called; if `value` is a `RawArray` the array is written to the selection in the destination array. A check is performed to ensure that the shape of the value array and destination in the `ZarrArray` are the same, and then `setRawArray` is called.

Finally, a signature with both a `chunk` of type `RawArray` and additional `chunkSelection` will call `setRawArrayFromChunkItem` quick indexes both the `chunk` and `destination` in tandem. I don't anticipate many users will use this final signature, but the former are analogous to what you had implemented earlier.
